arwen/host/execution.go

```diff
@@ -532,3 +568,336 @@ func (host *vmHost) createETHCallInput() []byte {
+}
+
+/**
+ * processCallbackStack is triggered when a callback was received from another host through a transaction.
+ *  It will return an error if we receive a callback and we don't have it's associated data in the storage.
+ *  If the associated callback was found in the pending set, it will be removed - It should not be executed
+ *   again since it was executed in the callSCMethod step
+ */
+func (host *vmHost) processCallbackStack() error {
+	runtime := host.Runtime()
+	storage := host.Storage()
+
+	storageKey, err := arwen.CustomStorageKey(host.Crypto(), arwen.AsyncDataPrefix, runtime.GetOriginalTxHash())
+	if err != nil {
+		return err
+	}
+
+	buff := storage.GetStorage(storageKey)
+
+	asyncInfo := &vmcommon.AsyncContextInfo{}
+	err = json.Unmarshal(buff, &asyncInfo)
+	if err != nil {
+		return err
+	}
+
+	vmInput := runtime.GetVMInput()
+	var asyncCallPosition int
+	var currentContextIdentifier string
+	for contextIdentifier, asyncContext := range asyncInfo.AsyncContextMap {
+		for position, asyncCall := range asyncContext.AsyncCalls {
+			if bytes.Equal(vmInput.CallerAddr, asyncCall.Destination) {
+				asyncCallPosition = position
+				currentContextIdentifier = contextIdentifier
+				break
+			}
+		}
+
+		if len(currentContextIdentifier) > 0 {
+			break
+		}
+	}
+
+	if len(currentContextIdentifier) == 0 {
+		return arwen.ErrCallBackFuncNotExpected
+	}
+
+	// Remove current async call from the pending list
+	currentContextCalls := asyncInfo.AsyncContextMap[currentContextIdentifier].AsyncCalls
+	currentContextCalls[asyncCallPosition] = currentContextCalls[len(currentContextCalls)-1]
+	currentContextCalls[len(currentContextCalls)-1] = nil
+	currentContextCalls = currentContextCalls[:len(currentContextCalls)-1]
+
+	if len(currentContextCalls) == 0 {
+		// call OUR callback for resolving a full context
+		delete(asyncInfo.AsyncContextMap, currentContextIdentifier)
+	}
+
+	// If we are still waiting for callbacks we return
+	if len(asyncInfo.AsyncContextMap) > 0 {
+		return nil
+	}
+
+	_, err = storage.SetStorage(storageKey, nil)
+	if err != nil {
+		return err
+	}
+
+	// Now figure out if we can execute the callback here or different shard
+	if !host.canExecuteSynchronouslyOnDest(asyncInfo.AsyncInitiator.CallerAddr) {
+		err = host.sendStorageCallbackToDestination(asyncInfo.AsyncInitiator)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// The caller is in the same shard, execute it's callback
+	callbackCallInput, err := host.createCallbackContractCallInput(
+		host.Output().GetVMOutput(),
+		asyncInfo.AsyncInitiator.CallerAddr,
+		arwen.CallbackDefault,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+
+	callbackVMOutput, _, callBackErr := host.ExecuteOnDestContext(callbackCallInput)
+	err = host.processCallbackVMOutput(callbackVMOutput, callBackErr)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/**
+ * setupAsyncCallsGasByPercentages takes the percentage of gas set up by the SC developer for each call
```

Comment:
wouldn't it be easier for developer to set exact gas value for each of these asyncCalls instead of percentage ?

Reply:
Agreed, absolute gas instead of percentage would be much better for the developer.

Reply:
Done